### PR TITLE
sql: Implement ALTER TABLE ... LOCALITY REGIONAL BY TABLE from GLOBAL

### LINKED ISSUE #58343: sql: Implement ALTER TABLE ... GLOBAL from all localities except REGIONAL BY ROW
As described in the functional specification.

### LINKED ISSUE #59249: sql: order show regions from database by primary region
It'd be nice if ordered the results in SHOW REGIONS FROM DATABASE by primary region to make it easier to read:
```SQL
root@127.0.0.1:26257/movr> SHOW REGIONS FROM DATABASE movr;
  database |    region    | primary |  zones
-----------+--------------+---------+----------
  movr     | europe-west1 |  false  | {b,c,d}
  movr     | us-east1     |  true   | {b,c,d}
  movr     | us-west1     |  false  | {a,b,c}
(3 rows)

Time: 14ms total (execution 13ms / network 0ms)
```

### PR DESCRIPTION
**Pull request by ajstorm**
Implements ALTER TABLE ... LOCALITY REGIONAL BY TABLE (all permutations)
from LOCALITY GLOBAL.

Release note (sql change): Implements ALTER TABLE ... LOCALITY REGIONAL
BY TABLE from LOCALITY GLOBAL.

Resolves #58343.
Also resolves #59249.

### CODE REVIEW

**Comment by ajstorm** [comment]
Resolves #58343

**Comment by ajstorm** [comment]
Also resolves #59249.

File: pkg/sql/alter_table_locality.go [hunk 1]
```diff
@@ -326,3 +268,75 @@ func (n *alterTableSetLocalityNode) startExec(params runParams) error {
...
+
+func generateNewLocalityConfigForLocalityGlobal() *descpb.TableDescriptor_LocalityConfig {
+	return &descpb.TableDescriptor_LocalityConfig{
+		Locality: &descpb.TableDescriptor_LocalityConfig_Global_{
+			Global: &descpb.TableDescriptor_LocalityConfig_Global{},
+		},
+	}
+}
+
+func generateNewLocalityConfigForLocalityRegionalByTable(
```
Comment (otan) [inline pkg/sql/alter_table_locality.go:280]: nit: i'd get rid of `generate` or `new` -- `generateLocalityConfigFor...` or `newLOcalityFor...`

File: pkg/sql/alter_table_locality.go [hunk 2]
```diff
@@ -70,72 +70,59 @@ func (n *alterTableSetLocalityNode) Next(runParams) (bool, error) { return false
...
-		tree.AsStringWithFQNames(&n.n, params.Ann()),
-	); err != nil {
+	return nil
+}
+
+func (n *alterTableSetLocalityNode) alterTableLocalityRegionalByTableToGlobal(
+	params runParams, dbDesc *dbdesc.Immutable, tableName tree.TableName,
+) error {
+	const operation string = "alter table locality REGIONAL BY TABLE to GLOBAL"
+	if err := assertIsMultiRegionDatabase(dbDesc, operation); err != nil {
```
Comment (otan) [inline pkg/sql/alter_table_locality.go:108]: can we do just the once at the startExec phase?

**Review by otan — APPROVED** [review]
> Resolves #58343
> Also resolves #59249.

put these as separate lines in the description, and github will close the issues when this merges

**Review by ajstorm — COMMENTED** [review]
Ahh, didn't realize it had to be in the commit message.  Changed.

TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-:-MS0WJfg1897lL6710G5:bkts68d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/alter_table_locality.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MRvXajqfAQw5_fmGVaW-r2-108:-MS0Wl7T67Is1KXf9bQs:b-4kx9zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4226b0c09ad409432e02575cd5fb52d1601b12ef/pkg/sql/alter_table_locality.go#L108)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can we do just the once at the startExec phase?
</blockquote></details>

I was worried about the helper functions being called from elsewhere, but considering that they're methods of the `alterTableSetLocalityNode`, that fear seems to be unwarranted.  I've removed the checks from the helpers.
___
*[pkg/sql/alter_table_locality.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MRvXajqfAQw5_fmGVaW-r2-280:-MS0XiIZ0LuoZSRni98M:bin91j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4226b0c09ad409432e02575cd5fb52d1601b12ef/pkg/sql/alter_table_locality.go#L280)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: i'd get rid of `generate` or `new` -- `generateLocalityConfigFor...` or `newLOcalityFor...`
</blockquote></details>

You're in luck - these helper methods will be gone once I rebase with @arulajmani's changes.

**Comment by otan** [comment]
not commit message, just PR description

(personal nit but putting it in the commit messages spams the issue with commit hashes)

**Comment by ajstorm** [comment]
Taking them out also spams the issue :-P

They're out now.  Will use PR technique going forward.

**Comment by ajstorm** [comment]
@arulajmani Merge is done now.  When you have time, PTAL.

**Review by arulajmani — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS56Him4XdyV6UswkLE:-MS56Him4XdyV6UswkLF:b-4rft1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L76)):*
> ```Go
> 	params runParams, dbDesc *dbdesc.Immutable, tableName tree.TableName,
> ) error {
> 	const operation string = "alter table locality GLOBAL to REGIONAL BY TABLE"
> ```

nit: we can just move this inline
___
*[pkg/sql/alter_table_locality.go, line 81 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS55J9S7q5tmSngY5zA:-MS55J9S7q5tmSngY5zB:b5toljz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L81)):*
> ```Go
> 			"invalid call %q on incorrect table locality. %v",
> 			operation,
> 			n.tableDesc.LocalityConfig,
> ```

I don't like the idea of printing the entire locality config here (which is quite unreadable because of how nested it is). Can we instead use the `FormatLocalityConfig` here? 
___
*[pkg/sql/alter_table_locality.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53pfw-0ytgkeV1hnE:-MS53pfw-0ytgkeV1hnF:b-7hq4xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L87)):*
> ```Go
> 	n.tableDesc.SetTableLocalityRegionalByTable(n.n.Locality.TableRegion)
> 
> 	// Finalize the alter by writing a new table descriptor and updating the zone configuration.
> ```

nit: could you wrap this comment at 80 chars please. 
___
*[pkg/sql/alter_table_locality.go, line 113 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS54LbqDGJ5oGASS1Pb:-MS54LbqDGJ5oGASS1Pc:bdzmday) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L113)):*
> ```Go
> 	n.tableDesc.SetTableLocalityGlobal()
> 
> 	// Finalize the alter by writing a new table descriptor and updating the zone configuration.
> ```

nit: 80 chars :sweat_smile: 
___
*[pkg/sql/alter_table_locality.go, line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS56cY02bichIURP5UC:-MS56cY02bichIURP5UD:b9p22mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L170)):*
> ```Go
> 
> 	// Get the fully resolved table name for use in the calls below.
> 	resolvedSchema, err := params.p.Descriptors().GetImmutableSchemaByID(
> ```

I like that we've moved this thing to one place, instead of duplicating it in a bunch of places as before. I think it'll be cleaner if we move this (and tableName below) to `validateAndWriteNewTableLocalityAndZoneConfig` instead. That way, we don't have to pass this around in all the locality change functions (and we're constructing this closer to its use).
___
*[pkg/sql/alter_table_locality.go, line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS52oaM30iZen3cEh4c:-MS52oaM30iZen3cEh4d:b4wltc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L249)):*
> ```Go
> }
> 
> // validateAndWriteNewTableLocalityAndZoneConfig validates the newly updated LocalityConfig
> ```

nit: could you wrap this comment to 80 chars please. 
___
*[pkg/sql/alter_table_locality.go, line 275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53-uEB3TJV1ReG_Vc:-MS53-uEB3TJV1ReG_Vd:b-9ddll4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L275)):*
> ```Go
> 
> 	// Update the zone configuration.
> 	if err := params.p.applyZoneConfigFromTableLocalityConfig(
> ```

nit: you can just return the function call here as opposed to `if err := ...`

**Review by ajstorm — COMMENTED** [review]
TFTR!  Changes made below

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-:-MS8T1uO8xvE7UUrDM0s:b5t1c5r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/alter_table_locality.go, line 76 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS56Him4XdyV6UswkLE:-MS8EITK1lYp6F0Qbn1o:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L76)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: we can just move this inline
</blockquote></details>

Done
___
*[pkg/sql/alter_table_locality.go, line 81 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS55J9S7q5tmSngY5zA:-MS8LyOp-kvpfihHcPlr:b7qg8x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't like the idea of printing the entire locality config here (which is quite unreadable because of how nested it is). Can we instead use the `FormatLocalityConfig` here? 
</blockquote></details>

I'm assuming you mean `FormatTableLocalityConfig`.  If so, Done.  If not :man_shrugging: 
___
*[pkg/sql/alter_table_locality.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53pfw-0ytgkeV1hnE:-MS8Mj_30XFZPGIlgA6y:bx9vrz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L87)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: could you wrap this comment at 80 chars please. 
</blockquote></details>

Happy to do this, but is this formalized anywhere?  I'm pretty sure I have many more comments that are longer than 80 chars.  I also just now went down a deep rabbit hole about standardization of line length.  While we do lint for line length (at 100 chars), I don't see any linting (or mention of linting) around comments.  In fact, internal documentation around commenting style presents examples where the comments are longer than 80 chars.

Again, no objection, but if we're going to suggest this in code reviews, we should write it down somewhere.  Maybe in the code commenting guidelines in Confluence?  That being said, that would only be visible to CRL members.  Ideally, we'd want this in the repo somewhere.

Oh, and BTW, done. :smile: 
___
*[pkg/sql/alter_table_locality.go, line 113 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS54LbqDGJ5oGASS1Pb:-MS8RIKjDV1Vpfe_jd-s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L113)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: 80 chars :sweat_smile: 
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_locality.go, line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS56cY02bichIURP5UC:-MS8RnP56vRA17mLSUyK:bik0ryq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L170)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I like that we've moved this thing to one place, instead of duplicating it in a bunch of places as before. I think it'll be cleaner if we move this (and tableName below) to `validateAndWriteNewTableLocalityAndZoneConfig` instead. That way, we don't have to pass this around in all the locality change functions (and we're constructing this closer to its use).
</blockquote></details>

Good call!  I refactored this code a bunch of times and didn't realize at the end that tableName only got used in `validate`.  Fixed.
___
*[pkg/sql/alter_table_locality.go, line 249 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS52oaM30iZen3cEh4c:-MS8S5hQ6yTKqUmsgHa2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L249)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: could you wrap this comment to 80 chars please. 
</blockquote></details>

Done.
___
*[pkg/sql/alter_table_locality.go, line 275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53-uEB3TJV1ReG_Vc:-MS8S8iTAwAmUA0Vn7mO:b33anmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L275)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: you can just return the function call here as opposed to `if err := ...` 
</blockquote></details>

I'm inclined to leave it as is.  What I worry about is if someone later comes along and has to make a call after the `applyZoneConfig` call, if they don't change the `return`, then their call will get silently skipped.  You'd think the compiler would catch this, but it's only a warning.

I think it's fine for stuff like `logEvent`, which should always be the last line of the function.

**Review by arulajmani — APPROVED** [review]
Thanks for making the changes! :shipit: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-:-MS8gh7x656p0rojI40S:bat6wqo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 81 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS55J9S7q5tmSngY5zA:-MS8dKjHBVb1lE-B-QJE:bzenwwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I'm assuming you mean `FormatTableLocalityConfig`.  If so, Done.  If not :man_shrugging: 
</blockquote></details>

Yeah that's what I meant, my bad on mistyping. 
___
*[pkg/sql/alter_table_locality.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53pfw-0ytgkeV1hnE:-MS8_1TP9_A4ItRJqGw5:b-ley4zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L87)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Happy to do this, but is this formalized anywhere?  I'm pretty sure I have many more comments that are longer than 80 chars.  I also just now went down a deep rabbit hole about standardization of line length.  While we do lint for line length (at 100 chars), I don't see any linting (or mention of linting) around comments.  In fact, internal documentation around commenting style presents examples where the comments are longer than 80 chars.

Again, no objection, but if we're going to suggest this in code reviews, we should write it down somewhere.  Maybe in the code commenting guidelines in Confluence?  That being said, that would only be visible to CRL members.  Ideally, we'd want this in the repo somewhere.

Oh, and BTW, done. :smile: 
</blockquote></details>

It's in the style guide here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+coding+guidelines#Gocodingguidelines-LineLengthO
___
*[pkg/sql/alter_table_locality.go, line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS56cY02bichIURP5UC:-MS8ef058WxmBRC2OIxm:b-j1we20) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L170)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Good call!  I refactored this code a bunch of times and didn't realize at the end that tableName only got used in `validate`.  Fixed.
</blockquote></details>

:raised_hands: 
___
*[pkg/sql/alter_table_locality.go, line 275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53-uEB3TJV1ReG_Vc:-MS8ewmGBY33PZHhCsgT:baexget) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L275)):*
> You'd think the compiler would catch this, but it's only a warning.

Doesn't  the IDE shout at you though? FWIW I think this is one of those things that we strive to do as much as possible in our code, anecdotally from a bunch of reviews I've gotten, but we sometimes things slip by. 
___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS8flH6DNggjHQYtSt5:bhyt00r) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
> ```Go
> 			// While we're in an error path and generally it's bad to return a
> 			// different error in an error path, we will only get an error here if the
> 			// locality is corrupted, in which case, it's probably the right error
> ```

Maybe this is for another PR, but I'm wondering if `FormatTableLocalityConfig` should return an error at all. Rather, this might just benefit from returning an `invalid locality` string instead if the table locality is of the wrong type (which shouldn't really happen, as validation of the table desc should ensure the locality is type is sane, if it doesn't do so already). Lemme know if you're on board with this and I'll make this change the next time I'm around here.

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53pfw-0ytgkeV1hnE:-MS8hJ8G6BIwby73mft9:bntl6sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L87)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

It's in the style guide here: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+coding+guidelines#Gocodingguidelines-LineLengthO
</blockquote></details>

TIL! :smile: 
___
*[pkg/sql/alter_table_locality.go, line 275 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS53-uEB3TJV1ReG_Vc:-MS8hCEc-IB4i8b2MZtM:b-9ar1ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b81d82279db630b5336ccc5a6d4ee35fe6f2290/pkg/sql/alter_table_locality.go#L275)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> You'd think the compiler would catch this, but it's only a warning.

Doesn't  the IDE shout at you though? FWIW I think this is one of those things that we strive to do as much as possible in our code, anecdotally from a bunch of reviews I've gotten, but we sometimes things slip by. 
</blockquote></details>

The IDE "shouts" by highlighting the call in yellow.  Very easy to miss.
___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS8hUSYDB-uDqHf0ruL:b-9y7yx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Maybe this is for another PR, but I'm wondering if `FormatTableLocalityConfig` should return an error at all. Rather, this might just benefit from returning an `invalid locality` string instead if the table locality is of the wrong type (which shouldn't really happen, as validation of the table desc should ensure the locality is type is sane, if it doesn't do so already). Lemme know if you're on board with this and I'll make this change the next time I'm around here. 
</blockquote></details>

I thought that too earlier today.  The problem is that in some of the cases where it's used, it's better to return the error than print "INVALID LOCALITY".  For instance, look at the call in `showCreateLocality()`.  There's also the call in `ValidateTableLocalityConfig()`.  That call does what you're suggesting, but I'd argue that's the wrong thing to do here as we're (somewhat) silently ignoring database corruption.  As I've done here, I think we need to return the right (and far worse) error here.  That being said, if you're looking for a follow-on project, I'd clean up the callers that are silently eating the error.

**Comment by ajstorm** [comment]
TFTR all.

bors r=arulajmani, otan

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS8rNMj15Wmkj1r3HIZ:b-4gyfgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

I thought that too earlier today.  The problem is that in some of the cases where it's used, it's better to return the error than print "INVALID LOCALITY".  For instance, look at the call in `showCreateLocality()`.  There's also the call in `ValidateTableLocalityConfig()`.  That call does what you're suggesting, but I'd argue that's the wrong thing to do here as we're (somewhat) silently ignoring database corruption.  As I've done here, I think we need to return the right (and far worse) error here.  That being said, if you're looking for a follow-on project, I'd clean up the callers that are silently eating the error.
</blockquote></details>

 > That call does what you're suggesting, but I'd argue that's the wrong thing to do here as we're (somewhat) silently ignoring database corruption.

It's not really database corruption though -- this can only happen if someone added a new type of locality config but forgot to add a case for it in the `FormatTableLocalityConfig` method.

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS8tEGBCReBqOCqDKyU:bxezotl) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

 > That call does what you're suggesting, but I'd argue that's the wrong thing to do here as we're (somewhat) silently ignoring database corruption.

It's not really database corruption though -- this can only happen if someone added a new type of locality config but forgot to add a case for it in the `FormatTableLocalityConfig` method. 
</blockquote></details>

True, that can also occur.  The case I'm more worried about though, is that the proto gets corrupted and we get some garbage value back here.  Isn't that possible as well?

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS8uhXr0ESG1eZ7zW6H:bq371pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

True, that can also occur.  The case I'm more worried about though, is that the proto gets corrupted and we get some garbage value back here.  Isn't that possible as well?
</blockquote></details>

Theoretically, yes, but there's more pressing concerns if storage gets corrupted. We have a replica consistency checker, which will crash the node if it detects that one of the replicas doesn't hash the same as other replicas of the same range.

**Comment by ajstorm** [comment]
Had to rebase.  Trying again...

bors r=arulajmani,otan

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59407)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 81 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/59407#-MS8flH6DNggjHQYtSt4:-MS90s0I8TtmTWSkOUhK:bchyqy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/815ff02905d69d80c2dbd1f3e854889b906ea644/pkg/sql/alter_table_locality.go#L81)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Theoretically, yes, but there's more pressing concerns if storage gets corrupted. We have a replica consistency checker, which will crash the node if it detects that one of the replicas doesn't hash the same as other replicas of the same range. 
</blockquote></details>

Let's bikeshed on Slack.  Corruption or no corruption, I think returning the error is better than silently mapping it to an "INVALID LOCALITY" string, which in theory could get missed by human readers, and even perhaps some testcases.

**Comment by ajstorm** [comment]
Merge problems fixed.  

bors r=arulajmani,otan

**Pull request merged**